README.txt
= pik

http://github.com/vertiginous/pik/tree/master
Gordon Thiesfeld

== DESCRIPTION:

  Pik is a tool to switch between multiple versions of ruby on Windows.  
  
  Or to put it another way
  
    <code>doskey rvm=pik $*</code>
  
  You have to tell it where your different ruby versions live using 
  'pik add'.  Then you can change to one by using 'pik switch'.
  
  It also has a "sort of" multiruby functionality in 'pik run'.
   

== FEATURES/PROBLEMS:

  Currently, changes are to the open cmd session only.  I haven't wired up the --global switch yet.

  Specs are very incomplete.

  'pik config' could be dangerous.  Use only if you know what you're doing.

  Only works on MRI at present.

== SYNOPSIS:

  <code>
  pik commands are:
  
      add           Add another ruby location to pik.
      checkup       Checks your environment for current Ruby best practices.
      list|ls       Lists ruby versions that pik is aware of.
      run           Runs command with all version of ruby that pik is aware of.
      remove|rm     Remove a ruby location from pik.
      switch|sw     Switches to another ruby location.
      help          Displays help topics.
  
  
  For help on a particular command, use 'pik help COMMAND'.
  </code>

  Example:

  <code>
    C:\>pik run "gem in hpricot --no-ri --no-rdoc"
     == Running with 185: ruby 1.8.5 (2006-12-25 patchlevel 12) [i386-mswin32] ==
    Successfully installed hpricot-0.8.1-x86-mswin32
    1 gem installed
    
     == Running with 186: ruby 1.8.6 (2008-08-11 patchlevel 287) [i386-mswin32] ==
    Successfully installed hpricot-0.8.1-x86-mswin32
    1 gem installed
    
     == Running with 186: ruby 1.8.6 (2009-03-31 patchlevel 368) [i386-mingw32] == 
    Building native extensions.  This could take a while...
    Successfully installed hpricot-0.8.1
    1 gem installed
    
     == Running with 191: ruby 1.9.1p0 (2009-01-30 revision 21907) [i386-mingw32] == 
    Building native extensions.  This could take a while...
    ERROR:  Error installing hpricot:
            ERROR: Failed to build gem native extension.
    
    ... Errors because I haven't added the devkit ...
    
     == Running with 191: ruby 1.9.1p129 (2009-05-12 revision 23412) [i386-mingw32] == 
    Building native extensions.  This could take a while...
    Successfully installed hpricot-0.8.1
    1 gem installed
    
    </code>
    
== REQUIREMENTS:

  Windows, Ruby, Rubygems

== INSTALL:

  1. gem install vertiginous-pik
  
  2. pik add path_to_ruby_bin # add your current ruby's bin dir
  
  3. repeat step 2 once for each version of ruby you have,
     or use 'pik add -i' for interactive console to do the same.
  
  4. pik run "gem install vertiginous-pik" 

== LICENSE:

(The MIT License)

Copyright (c) 2009 Gordon Thiesfeld

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.2.8 / 2010-06-22
** BACKWARD COMPATIBILITY CHANGE 
  pik will now install new rubies to %USERPROFILE%\rubies\, instead of %PIK_BIN%\pik\  this was necessary to all for installation into %PROGRAMFILES% dir.  If you'd like to keep the old path, run:
    >pik config install_dir=C:\path\to\pik
**
* fixed bug in 'devkit update' - gcc.bat file was writing to/overwriting sh.bat
* added support for powershell
* added support for Git Bash
* it is now possible to install pik to %PROGRAMFILES%

=== 0.2.7 / 2010-06-15
* cleaner list output
* pik switch/use is no longer case sensitive
    '>pik iron' vice '>pik Iron'
* created devkit implmentation (pik install devkit)
* created devkit update command (pik devkit update) creates batch files if you install
    a newer version of ruby after devkit.
* added devkit config option, if you already have it installed:
    >pik config devkit=c:\ruby\devkit
    >pik devkit update
* fixed Implementations::IronRuby to catch newer versions, and ignore ipy packages
* improved pik_install instructions
* pik_install displays usage message when no argument is given
* fixed some bugs in info command
* fixed pik.bat to run in tcc/le
* removed RubyGems dependency ( the executable was causing weird errors with plugins)
* fixed Implementations::Ruby to catch RC1 7z files
* fixed regression in version option

=== 0.2.6 / 2009-11-08

* deprecated checkup command in favor of info command
* added info command
* added benchmark command
* fixes to gemsync command
* don't need to be admin to run pik default
* added update command
* gemsync command - platform aware - remote option - quiet option


=== 0.2.5 / 2009-10-29

* use upx (http://upx.sf.net/) to compress pik_runner.exe
* fixes uninstall story so it doesn't conflict with other stories
* bugfix in uninstall command
* made all prompts default to 'yes'.  bugfix:interactive add would prompt for jruby\bin dirs twice because of the existence of jruby.exe and jruby.bat
* fixed downloading of seven_zip error
* uninstall command
* fixes bug in BatchFileEditor#switch_path_to
* switched to HPricot for package searches, requires rubygems and exerb-mingw - closes gh-30

=== 0.2.4 / 2009-10-28

* added rake command
* added cucumber story for rake command, refactor of run path switching, added irake.bat as a rake executable for ironruby
* fix to compile zlib1.dll into pik_runner.exe.  closes gh-29 Thanks, Luis Lavena and, Roger Pack
* fixed homepage in gemspec for gemcutter.org

=== 0.2.3 / 2009-10-25

* install command improvements - better error message on unrecognized implementation, better version picking, better directory naming, renamed some variables to be more concise (not everything is a version - implementations have versions, and versions have packages)
* added remote option to list command
* better error message when a version was deleted, but not removed from the config.  Thanks Roger Pack
* bugfix: run/gem/ruby command hangs when waiting for input. Thanks Roger Pack
* renamed tools/pik.exe to pik_runner.exe to minimize confusion/conflicts. Thanks Luis Lavena
* Upgraded to exerb 0.4.4.  Thanks Alexey Borzenkov
* Added documentation for tags commands. Thanks Luis Lavena
* 'pik tags' with no options will display the help docs, instead of throwing a cryptic error.  
* Added docs for ruby and gem command. Improved docs for run command

=== 0.2.2 / 2009-10-20

* Pik::Command#get_version now handles paths with spaces.  Thanks to Dominic Sisneros and Vassilis Rizopoulos 
* All Cucumber scenarios are run against 'C:\temp\path with spaces' instead of 'C:\temp'
* Added :rebuild task as a dependency to :package task, so that the executable gets rebuilt on packaging.  Thanks, skim.

=== 0.2.1 / 2009-10-19

* bugfix - path to pik could have spaces - thanks, Dominic Sisneros
* update install instructions

=== 0.2.0 / 2009-10-18

* pik is now an exerb compiled executable, and must be installed to a location that is in the system path, but is
  seperate from ruby
* added pik_install command
* added support for IronRuby and Jruby
* added install command to install new Ruby versions
* added tag and tags command
* added cucumber functional tests
* updated gem post-install message
* changed pik run to run in ruby instead of in a batch file.  
* Added pik ruby and pik gem commands.

=== 0.1.1 / 2009-09-25

* add rubyforge info to README
* Fixes bug in add command that wouldn't allow you to add - GH-13

=== 0.1.0 / 2009-09-24

* removed jruby support for the time being (lack of win32ole is a problem)
* added more helpful messages to pik add
* renamed gemdup to gemsync.  fixed the code up a bit
* added default and implode commands, shortcuts for switch, more specs, lots of refactors
* removed add_gem_home command.  GEM_HOME can be configured with pik config
* change remove 'force' to 'quiet'
* added simpler switch semantics (`pik 191 p243` instead of `pik switch 191 p243`)
* added gemdup command
* option parsing is now done with optparse and not a regex
* rewrote pik to use classes for each command in the hopes that my code would be easier to read
* Added post install instructions - issue 5. Switched to Hoe.spec for Hoe compatibility.  added Hoe-git plugin.
* added verbose output and current version specifier to 'list'
* added 'use' as an alias for 'switch'
* cleaner method of remvoving .pik/pik batch file, fixed some long lines

=== 0.0.3 / 2009-06-10

* added 'add_gem_home' command
  due to changes in the config, you'll have to delete your "%HOME%\.pik\config.yml" and recreate it.
* don't use GEM_HOME on XP with rubygems 1.3.4, wait for 1.3.5, or point them to a path without spaces.

=== 0.0.2 / 2009-06-02
 
* several minor enhancements
  *updated help messages
  *updated readme
  *some formatting of output

=== 0.0.1 / 2009-06-02

* 1 major enhancement

  * Birthday!



features/env.rb
config=<<CONFIG
--- 
"092: IronRuby 0.9.2.0 on .NET 2.0.0.0": 
  :path: !ruby/object:Pathname 
    path: C:/temp/more spaces in path/ruby/IronRuby-092/bin
"151: jruby 1.5.1 (ruby 1.8.7 patchlevel 249) (2010-06-06 f3a3480) (Java HotSpot(TM) Client VM 1.6.0_18) [x86-java]": 
  :path: !ruby/object:Pathname 
    path: C:/temp/more spaces in path/ruby/JRuby-151/bin
"186: ruby 1.8.6 (2010-02-04 patchlevel 398) [i386-mingw32]": 
  :path: !ruby/object:Pathname 
    path: C:/temp/more spaces in path/ruby/Ruby-186-p398-2/bin
"187-test: ruby 1.8.7 (2010-01-10 patchlevel 249) [i386-mingw32]": 
  :path: !ruby/object:Pathname 
    path: C:/temp/more spaces in path/ruby/Ruby-187-p249-1/bin
"187: ruby 1.8.7 (2010-01-10 patchlevel 249) [i386-mingw32]": 
  :path: !ruby/object:Pathname 
    path: C:/Ruby/ruby-187-p249/bin
"191: ruby 1.9.1p378 (2010-01-10 revision 26273) [i386-mingw32]": 
  :path: !ruby/object:Pathname 
    path: C:/temp/more spaces in path/ruby/Ruby-191-p378-1/bin
--- 
:download_dir: !ruby/object:Pathname 
  path: C:\\temp\\path with spaces\\downloads
:install_dir: !ruby/object:Pathname 
  path: C:\\temp\\path with spaces
CONFIG

require 'yaml'
require 'pathname'
require 'fileutils'
require 'lib/pik'
require 'rbconfig'

REAL_PATH  = SearchPath.new(ENV['PATH']).replace(RbConfig::CONFIG['bindir'], 'C:\temp\more spaces in path\ruby\Ruby-186-p398-2\bin').join
OTHER_RUBY  = 'c:\\temp\\more spaces in path\\ruby\\Ruby-191-p378-1\\bin'

ENV['HOME'] = "C:\\temp\\path with spaces"

PIK_LOG = 'log\\output.log'
TEST_PIK_HOME  = Pathname.new( ENV['HOME'] || ENV['USERPROFILE'] ) + '.pik'
FAKE_PIK_HOME = 'c:/temp/path with spaces/.pik'

Before do
  ENV['PATH'] = REAL_PATH
  FileUtils.rm_rf FAKE_PIK_HOME
  FileUtils.mkdir_p FAKE_PIK_HOME
  File.open(File.join(FAKE_PIK_HOME, 'config.yml'), 'w'){|f| f.puts config }
end

